compiler/src/dotty/tools/dotc/typer/Applications.scala

```diff
@@ -529,50 +529,86 @@ trait Applications extends Compatibility {
-      }
-
-      def handlePositional(pnames: List[Name], args: List[Trees.Tree[T]]): List[Trees.Tree[T]] =
-        args match {
-          case (arg: NamedArg @unchecked) :: _ =>
-            val nameAssocs = for (case arg @ NamedArg(name, _) <- args) yield (name, arg)
-            handleNamed(pnames, args, nameAssocs.toMap, toDrop = Set(), missingArgs = false)
+      def handleNamed(names: List[(Name, Option[(Name, Annotation)])], args: List[Trees.Tree[T]],
+                      nameToArg: Map[Name, Trees.NamedArg[T]], toDrop: Set[Name], missingArgs: Boolean): List[Trees.Tree[T]] =
+
+        /** Returns the corresponding name parameter with the name that helped find it */
+        def findNameArgument(pname: (Name, Option[(Name, Annotation)])): Option[(Name, Trees.NamedArg[T])] =
+          pname._2 match
+            // This parameter has a deprecated name, looking for this first
+            case Some(name, annot) => nameToArg.get(name) match
+              // Argument with a deprecated name was found: emit a warning and change the parameter name
+              case Some(arg) =>
+                val sinceMsg = annot.argumentConstantString(1).map(version => s" (since: $version)").getOrElse("")
+                report.deprecationWarning(em"naming parameter $name is deprecated$sinceMsg", arg.srcPos)
+                Some((name, arg))
+              // fallback to the primary name
+              case None => nameToArg.get(pname._1).map((pname._1, _))
+            // no-deprecated name, we can look for the argument with the main name
+            case None => nameToArg.get(pname._1).map((pname._1, _))
+        end findNameArgument
+
+        def namesRest = if names.isEmpty then Nil else names.tail
+
+        names match
+        case pname :: pnames  if findNameArgument(pname).nonEmpty => findNameArgument(pname) match
+            case Some((name, arg)) => // named argument was found, move on to the next parameter
+              val namesToDrop = pname._2.map(_._1).toSet + pname._1 // drop both the deprecated and the primary name
+              arg :: handleNamed(pnames, args, nameToArg - name, toDrop ++ namesToDrop, missingArgs)
+            case _ => throw new IllegalStateException
+        case _ => args match
+          case (arg @ NamedArg(name, _)) :: args1 =>
+            if toDrop.contains(name) then // argument is already passed
+              handleNamed(names, args1, nameToArg, toDrop - name, missingArgs)
+            else if (nameToArg.contains(name) && names.nonEmpty) // argument is missing, pass an empty tree
+              genericEmptyTree :: handleNamed(names.tail, args, nameToArg, toDrop, missingArgs = true)
+            else // name not (or no longer) available for named arg
+              def msg =
+                if methodType.paramNames.contains(name) then
+                  em"parameter $name of $methString is already instantiated"
+                else
+                  em"$methString does not have a parameter $name"
```

Comment:
I had another idea (not what we discussed this afternoon). We could keep the logic as before but at this point, instead of failing, we do an addiitonal check. Say our argument is `x = e`. We check whether the method symbol has a `deprecatedName("x")` annotation on a parameter with new name `y`. If that's the case we change the `NamedArg` name to `y` and try again. If that succeeds with issue the deprecate name warning.


If that's the case, we try again with the